File: DSA/src/Questions/BinaryTree/Diameter.java
package Questions.BinaryTree;

public class Diameter {

    public int maxDepth(TreeNode root, int[] d) {
        if (root == null) {
            return 0;
        }
        int ld = maxDepth(root.left, d);
        int rd = maxDepth(root.right, d);
        d[0] = Math.max(d[0], (ld + rd));
        return 1 + Math.max(ld, rd);
    }

    public int diameterOfBinaryTree(TreeNode root) {
        int[] d = new int[1];
        maxDepth(root, d);
        return d[0];
    }

    public static void main(String[] args) {
        TreeNode root = new TreeNode(1);
        TreeNode right = new TreeNode(2);
        TreeNode rightLeft = new TreeNode(3);
        root.right = right;
        right.left = rightLeft;
        Diameter diameter = new Diameter();
        int result = diameter.diameterOfBinaryTree(root);
        System.out.println("Diameter of the tree: " + result);
    }
}

File: DSA/Python/Questions/BinaryTree/Diameter.py
"""
# Question: Diameter of Binary Tree
# Link: https://leetcode.com/problems/diameter-of-binary-tree/

# Time Complexity: O(N)
# Space Complexity: O(H)

# Algorithm:
# 1. DFS traversal
# 2. Track max diameter
# 3. Return height at each node
"""


class TreeNode:
    def __init__(self, val=0):
        self.val = val
        self.left = None
        self.right = None


class Diameter:
    def diameter_of_binary_tree(self, root: TreeNode) -> int:
        self.diameter = 0

        def height(node: TreeNode) -> int:
            if not node:
                return 0

            left_height = height(node.left)
            right_height = height(node.right)

            self.diameter = max(self.diameter, left_height + right_height)
            return max(left_height, right_height) + 1

        height(root)
        return self.diameter


def main():
    solution = Diameter()
    root = TreeNode(1)
    root.left = TreeNode(2)
    root.right = TreeNode(3)
    print(solution.diameter_of_binary_tree(root))


if __name__ == "__main__":
    main()
